### PR TITLE
Allow setting index_type (e.g., 'HASH' or 'BTREE')

### PR DESCRIPTION
**Pull request by ryanpbrewster**
May address https://github.com/JetBrains/Exposed/issues/895

### CODE REVIEW

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/Table.kt [hunk 1]
```diff
@@ -882,6 +882,17 @@ open class Table(name: String = "") : ColumnSet(), DdlAware {
...
     }
 
+    /**
+     * Creates an index with a custom index type.
+     *
+     * @param customIndexName Name of the index.
+     * @param columns Columns that compose the index.
+     * @param isUnique Whether the index is unique or not.
+     */
+    fun index(customIndexName: String? = null, indexType: String, vararg columns: Column<*>) {
```
Comment (Tapac) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/Table.kt:892]: Maybe it's better to extend existing `index` function with `indexType: String? = null`. It should not break existing code. 
But maybe it's not possible to create unique index with custom type?

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/vendors/MariaDBDialect.kt [hunk 2]
```diff
@@ -28,6 +25,10 @@ class MariaDBDialect : MysqlDialect() {
     override val functionProvider: FunctionProvider = MariaDBFunctionProvider
     override val supportsOnlyIdentifiersInGeneratedKeys: Boolean = true
 
+    override fun createIndex(index: Index): String {
```
Comment (Tapac) [inline exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/vendors/MariaDBDialect.kt:28]: Looks unnecessary,

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/Table.kt (hunk 1 above)
Comment (ryanpbrewster) [reply exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/Table.kt:892]: The limitations of custom index types are not very well standardized.
 - Postgres only supports unique indexes for `BTREE`
 - MySQL supports unique indexes using `HASH` and `BTREE`, but multi-column indexes must be `BTREE`
 - SQLite does not have any custom index types

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/vendors/MariaDBDialect.kt (hunk 2 above)
Comment (ryanpbrewster) [reply exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/vendors/MariaDBDialect.kt:28]: removed

File: exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/Table.kt (hunk 1 above)
Comment (ryanpbrewster) [reply exposed-core/src/main/kotlin/org/jetbrains/exposed/sql/Table.kt:892]: I went ahead and addedthis to the existing `index` function.

I do not believe there is a clean, vendor-agnostic API that describes exactly when custom index types are available. Perhaps it is best to provide the option and fail at runtime if the database implementation does not support the specified index.

**Comment by Tapac** [comment]
@ryanpbrewster , please fix test compilation and I'll merge PR.

**Comment by ryanpbrewster** [comment]
Updated, I believe the tests compile and pass now.

As a side note, I found it a bit tricky to get the tests to compile and run. I'm developing on a Mac, and it seems that the mysql test harness assumes a Linux distro. When I booted up an Ubuntu container and mounted the directory, I still ran into several issues with tests failing at master (several due to timestamp resolution issues, where my system seems to be rounding timestamps to millisecond resolution).

Is there any kind of contributor guide to getting set up to build and run tests?

**Comment by Tapac** [comment]
@ryanpbrewster , it's a known issue with embedded mysql (mxj) which should be replaced with test-containers. 

timestamp resolution was broken when Java9+ used, but it should be fixed at master already. If you'll face them after updating to master please file an issue.

**Pull request merged**